src/gin/nccl_ofi_gin.cpp

```diff
@@ -0,0 +1,734 @@
+/*
+ * Copyright (c) 2025 Amazon.com, Inc. or its affiliates. All rights reserved.
+ */
+
+#include "config.h"
+
+#include <gdrapi.h>
+
+#include "nccl_ofi.h"
+#include "nccl_ofi_api.h"
+#include "nccl_ofi_assert.h"
+#include "nccl_ofi_cuda.h"
+#include "nccl_ofi_rdma.h"
+
+#include "gin/nccl_ofi_gin.h"
+#include "gin/nccl_ofi_gin_reqs.h"
+
+/**
+ * The highest value of NSEG is used to flag an ack message
+ *
+ * TODO something better?
+ */
+#define WRITEDATA_ACK_NSEG ((1 << NUM_NUM_SEG_BITS) - 1)
+
+
+struct rdma_gin_connect_handle
+{
+	/* Number of rails */
+	uint16_t num_rails;
+	uint16_t num_control_rails;
+
+	/* A comm identitifer that uniquely identifies the comm on the sender
+	   side. The receiver must use this ID when sending messages to sender */
+	uint32_t comm_id;
+
+	/* Arrays of `MAX_NUM_RAILS` `nccl_ofi_rdma_ep_name_t`
+	 * structs. The member `num_rails` and `num_control_rails` indicate
+	 * the number of entries that are in use. */
+	nccl_ofi_rdma_ep_name_t control_ep_names[MAX_NUM_RAILS];
+	nccl_ofi_rdma_ep_name_t ep_names[MAX_NUM_RAILS];
+
+	/* Flush addr and its mr_key */
+	uint64_t flush_buff_addr;
+	uint64_t flush_buff_mr_key[MAX_NUM_RAILS];
+};
+
+
+static inline void set_flush_buff_info(nccl_net_ofi_rdma_domain_t *domain,
+				       rdma_gin_connect_handle &handle)
+{
+	auto &flush_buff = domain->flush_buff;
+	handle.flush_buff_addr = reinterpret_cast<uint64_t>(flush_buff.buffer);
+	/* Set mr key for first rail */
+	for (int i = 0; i < domain->num_rails; ++i) {
+		uint64_t key = fi_mr_key(flush_buff.mr_handle->mr[i].get());
+		assert_always(key != FI_KEY_NOTAVAIL);
+		handle.flush_buff_mr_key[i] = key;
+	}
+}
+
+
+static inline int rail_addr_insert(nccl_net_ofi_ep_rail_t &rail, const nccl_ofi_rdma_ep_name_t &ep_addr,
+				   int peer_rank, fi_addr_t &ofi_addr,
+				   std::unordered_map<fi_addr_t, uint64_t> &rank_map)
+{
+	int ret = fi_av_insert(rail.av.get(), ep_addr.ep_name, 1,
+				&ofi_addr, 0, nullptr);
+	if (ret != 1) {
+		NCCL_OFI_WARN("Failed to insert address for peer rank %d rail %hu", peer_rank,
+			      rail.rail_id);
+		return -EIO;
+	}
+	ret = 0;
+
+	auto res = rank_map.insert(std::make_pair(ofi_addr, peer_rank));
+	if (res.second == false) {
+		NCCL_OFI_WARN("Invalid duplicate address %lu for peer rank %d",
+				ofi_addr, peer_rank);
+		throw std::runtime_error("Invalid duplicate address");
```

Comment:
The C++ exception throwing here is inconsistent with the previous error handling.